packages/payment-proxy-client/src/file.ts

```diff
@@ -1,29 +1,158 @@
 import { NitroRpcClient } from "@statechannels/nitro-rpc-client";
+import { Voucher } from "@statechannels/nitro-rpc-client/src/types";
 
 export async function fetchFile(
-  baseUrl: string,
-  costPerByte: number,
-  fileSize: number,
+  url: string,
+  paymentAmount: number,
   selectedChannel: string,
   nitroClient: NitroRpcClient
 ): Promise<File> {
   const voucher = await nitroClient.CreateVoucher(
     selectedChannel,
-    fileSize * costPerByte
+    paymentAmount
   );
 
-  const response = await fetch(
-    `${baseUrl}?channelId=${voucher.ChannelId}&amount=${voucher.Amount}&signature=${voucher.Signature}`
+  const response = await fetch(addVoucherToUrl(url, voucher));
+
+  const fileName = parseFileNameFromUrl(response.url);
+
+  return new File([await response.blob()], fileName);
+}
+export async function fetchFileInChunks(
+  chunkSize: number,
+  url: string,
+  costPerByte: number,
+  selectedChannel: string,
+  nitroClient: NitroRpcClient,
+  updateProgress: (progress: number) => void
+): Promise<File> {
+  updateProgress(0);
+
+  const firstChunk = await fetchChunk(
```

Comment:
The HTTP range response will include a `Content-Range` header that includes the size of the total file. So after we get the first chunk we know the total length of the file we're requesting, as thus our progress as we download it.

If the size of the first chunk is bigger than the whole file, we just end up receiving the whole first in the first request.

Reply:
Does this play into the problem of knowing how much the download is going to cost ahead of time?

Reply:
It definitely helps! After we get the first response back we can parse the `Content-Range` header, and know the exact size of the file and how much it will cost (assuming you know the cost per byte being charged). We could probably even expand this by making a request for `0` bytes initially just to get the size of the file, so we know the cost in advance.

All of this relies on the use of the [HTTP range header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range). If a server doesn't support range headers we won't be able to do any of this, we can just request the whole file in one chunk.